polkadot/xcm/pallet-xcm/src/ethereum.rs

```diff
@@ -0,0 +1,191 @@
+		// Ensure all assets (including fees) have same reserve location.
+		ensure!(assets_transfer_type == fees_transfer_type, Error::<T>::TooManyReserves);
+		let (local_xcm, remote_xcm) = Self::build_ethereum_xcm_transfer(
+			origin.clone(),
+			dest.clone(),
+			beneficiary,
+			assets,
+			assets_transfer_type,
+			fees.clone().into_inner(),
+			weight_limit,
+		)?;
+
+		Self::execute_ethereum_xcm_transfer(origin, dest, fees, local_xcm, remote_xcm)
+	}
+
+	fn find_ethereum_assets_transfer_types(
+		assets: &[Asset],
+		fee: &Asset,
+		dest: &Location,
+	) -> Result<(TransferType, TransferType), Error<T>> {
+		let mut assets_transfer_type = None;
+		let fee_transfer_type =
+			T::XcmExecutor::determine_for(fee, dest).map_err(Error::<T>::from)?;
+
+		for (_, asset) in assets.iter().enumerate() {
+			if let Fungible(x) = asset.fun {
+				// If fungible asset, ensure non-zero amount.
+				ensure!(!x.is_zero(), Error::<T>::Empty);
+			}
+			let transfer_type =
+				T::XcmExecutor::determine_for(&asset, dest).map_err(Error::<T>::from)?;
+			if let Some(existing) = assets_transfer_type.as_ref() {
+				// Ensure transfer for multiple assets uses same transfer type (only fee may
+				// have different transfer type/path)
+				ensure!(existing == &transfer_type, Error::<T>::TooManyReserves);
+			} else {
+				// asset reserve identified
+				assets_transfer_type = Some(transfer_type);
+			}
+		}
+		Ok((fee_transfer_type, assets_transfer_type.ok_or(Error::<T>::Empty)?))
+	}
+
+	fn build_ethereum_xcm_transfer(
+		origin: Location,
+		dest: Location,
+		beneficiary: Location,
+		assets: Vec<Asset>,
+		transfer_type: TransferType,
+		fees: Vec<Asset>,
+		weight_limit: WeightLimit,
+	) -> Result<(Xcm<<T as Config>::RuntimeCall>, Option<Xcm<()>>), Error<T>> {
+		log::debug!(
+			target: "xcm::pallet_xcm::build_xcm_transfer_type",
+			"origin {:?}, dest {:?}, beneficiary {:?}, assets {:?}, transfer_type {:?}, \
+			fees_handling {:?}, weight_limit: {:?}",
+			origin, dest, beneficiary, assets, transfer_type, fees, weight_limit,
+		);
+		let fee_on_bridgehub = &fees[0];
+		let fee_on_ethereum = &fees[1];
+		let (local, mut remote) = match transfer_type {
+			TransferType::LocalReserve => {
+				let (local, remote) = Self::local_reserve_transfer_programs(
+					origin.clone(),
+					dest.clone(),
+					beneficiary,
+					assets,
+					FeesHandling::Batched { fees: fee_on_ethereum.clone() },
+					weight_limit,
+				)?;
+				Some((local, remote))
+			},
+			TransferType::DestinationReserve => {
+				let (local, remote) = Self::destination_reserve_transfer_programs(
+					origin.clone(),
+					dest.clone(),
+					beneficiary,
+					assets,
+					FeesHandling::Batched { fees: fee_on_ethereum.clone() },
+					weight_limit,
+				)?;
+				Some((local, remote))
+			},
+			_ => None,
+		}
+		.ok_or(Error::InvalidAssetUnsupportedReserve)?;
+		remote.inner_mut().push(BurnAsset(fee_on_bridgehub.clone().into()));
```

Comment:
Add `BurnAsset` into remote _xcm so that we can verify against the local cost on BH.

Reply:
Can you explain more what this `BurnAsset` is for?

the message arriving and executed on BridgeHub is:
```rust
	Xcm(vec![
		WithdrawAsset(Assets::from(vec![fee.clone()])),
		BuyExecution { fees: fee, weight_limit: Unlimited },
		ExportMessage {
			network: Ethereum { chain_id: ethereum_chain_id },
			destination: Here,
			xcm: inner_xcm,
		},
	]);
```
Where `fee` in `BuyExecution { fees: fee, weight_limit: Unlimited }` is `DOT`, no?
But this ^ is just for paying for `ExportMessage` which probably doesn't include relayer/ethereum tx cost, yes?

The `inner_xcm` is what `pallet-xcm` builds. And it uses WETH in inner `BuyExecution`? This `WETH` is practically the price for relayer/ethereum tx, no?

Why do you also want `BurnAsset(fee_on_bridgehub)` in the inner message? What does this `fee_on_bridgehub` (in DOT) represent?

Reply:
I've made `XcmRouter` on AH to use `UnpaidRemoteExporter` for bridging to Ethereum so there is no `BuyExecution` outside. 
In our outbound router https://github.com/Snowfork/polkadot-sdk/pull/135/files#diff-0bd5b75b88aee6992e3ddba0de8bd08d2ab07ecd38fc745c7d76daff56bba525

The `inner_xcm` contains a `BuyExecution` with the fee in `WETH`, we also need another instruction that contains the fee in DOT for the execution cost on BH, that's why I add the `BurnAsset(fee_on_bridgehub)`



Be more specific, the xcm arrives to BH is:

```
Xcm(vec![
        UnpaidExecution,
        ExportMessage {
            network: Ethereum { chain_id: ethereum_chain_id },
            destination: Here,
            xcm: Xcm(vec![
                WithdrawAsset(WETH,weth_amount), 
                ClearOrigin, 
                BuyExecution { fees: (WETH,weth_fee_amount)}, weight_limit: Unlimited }, 
                DepositAsset { assets: Wild(AllCounted(1))} }, 
                BurnAsset(Assets([Asset { id: AssetId(Location { parents: 1, interior: Here }), fun: Fungible(dot_fee_amount) }])), 
                SetTopic(),
        },
    ]);
```
As you may notice the `BuyExecution` inner xcm in `ExportMessage` will 

For the detailed xcm generated, just run the `cargo test -p bridge-hub-rococo-integration-tests --lib tests::snowbridge::send_weth_asset_from_asset_hub_to_ethereum  -- --nocapture` to check the log.